### PR TITLE
fix: Process was not started by this object

### LINKED ISSUE #5362: Process was not started by this object, so requested information cannot be determined.
**Do you want to request a *feature* or report a *bug*?**
:bug:

**What is the current behavior?**

```
System.InvalidOperationException
  HResult=0x80131509
  Message=Process was not started by this object, so requested information cannot be determined.
  Source=System
Callstack:
   at System.Diagnostics.Process.EnsureState(State state)
   at System.Diagnostics.Process.get_ExitCode()
   at GitCommands.Executable.ProcessWrapper.OnProcessExit(Object sender, EventArgs eventArgs) in D:\projekty\GitExt\GitCommands\Git\Executable.cs:line 109
   at System.Diagnostics.Process.OnExited()
   at System.Diagnostics.Process.RaiseOnExited()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading._ThreadPoolWaitOrTimerCallback.PerformWaitOrTimerCallback(Object state, Boolean timedOut)
```

**Environment you encounter the issue:**

- GitExtensions version: 3172d1c318460f5909a597cdea1d705b942a4ed1

### PR DESCRIPTION
**Pull request by RussKie**
In some circumstances GE may crash with `System.InvalidOperationException: Process was not started by this object, so requested information cannot be determined.`

The most likely scenario is Thread1 can pass the `if (_disposed == 0)` (line 107) and then Thread2 can come into play and call `Dispose` until the line 191. After that Thread1 gets resumed and works in the disposed state.

Closes #5362

### CODE REVIEW

**Comment by RussKie** [comment]
Should we assert that the disposal is

**Comment by RussKie** [comment]
Reworked it to use a lock

File: GitCommands/Git/Executable.cs [hunk 1]
```diff
@@ -91,22 +90,23 @@ public ProcessWrapper(string fileName, string arguments, string workDir, bool cr
...
 
             private void OnProcessExit(object sender, EventArgs eventArgs)
             {
                 // The Exited event can be raised after the process is disposed, however
                 // if the Process is disposed then reading ExitCode will throw.
-                if (_disposed == 0)
+                lock (_syncRoot)
                 {
-                    ExitCode = _process.ExitCode;
+                    if (_disposed == 0)
```
Comment (jbialobr) [inline GitCommands/Git/Executable.cs]: `_disposed` can be now a regular boolean.

File: GitCommands/Git/Executable.cs [hunk 2]
```diff
@@ -177,15 +177,18 @@ public StreamReader StandardError
...
             public void Dispose()
             {
-                if (Interlocked.CompareExchange(ref _disposed, 1, 0) != 0)
+                lock (_syncRoot)
                 {
-                    return;
+                    if (_disposed == 0)
+                    {
+                        _disposed = 1;
+                        return;
```
Comment (jbialobr) [inline GitCommands/Git/Executable.cs]: `return` should be in the `else` branch of this condition.

**Comment by jbialobr** [comment]
The `WaitForExit` also bothers me. It is strange that we can wait for a disposed object. It returns `ExitCode` in such a case, but the `ExitCode` may be undefined at that moment. I would rather throw an ObjectDisposed exception instead.

**Comment by RussKie** [comment]
/facepalm... indeed

File: GitCommands/Git/Executable.cs [hunk 3]
```diff
@@ -91,22 +90,23 @@ public ProcessWrapper(string fileName, string arguments, string workDir, bool cr
                     }
                 };
 
-                _logOperation = CommandLog.LogProcessStart(fileName, arguments);
-
                 _process.Exited += OnProcessExit;
-
                 _process.Start();
 
+                _logOperation = CommandLog.LogProcessStart(fileName, arguments);
```
Comment (drewnoakes) [inline GitCommands/Git/Executable.cs]: I'd avoid moving this line down. It exposes a different race where `OnProcessExit` tries to use a `_logOperation` before it's defined.

File: GitCommands/Git/Executable.cs [hunk 4]
```diff
@@ -177,15 +177,18 @@ public StreamReader StandardError
...
                 {
-                    return;
+                    if (_disposed == 0)
+                    {
+                        _disposed = 1;
+                        return;
+                    }
                 }
 
-                _process.Exited -= OnProcessExit;
```
Comment (drewnoakes) [inline GitCommands/Git/Executable.cs]: Ditto. I know it looks cleaner but I spent a lot of time tweaking the order of things here. Can't remember if this particular case is important, but if you only changed it to look nicer then consider reverting this reordering.

**Comment by drewnoakes** [comment]
> The WaitForExit also bothers me

Yes, there is a race in this case too. Take a look now.

**Pull request closed without merging**